examples/gaussian_process/Main.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE FunctionalDependencies #-}
+
+module Main where
+
+import Control.Monad (foldM)
+
+import Prelude hiding (exp)
+
+import Torch.Tensor
+import Torch.DType (DType (Float))
+import Torch.TensorFactories (eye', ones', rand', randn', zeros')
+import Torch.Functions
+import Torch.Autograd
+import Torch.NN
+
+-- | construct pairs of points on the axis
+makeAxis :: [Float] -> [Float] -> (Tensor, Tensor)
+makeAxis axis1 axis2 = (t, t')
+    where 
+        t = asTensor (fst <$> rngPairs)
+        t' = asTensor (snd <$> rngPairs)
+        pairs axis1' axis2' = [(t, t') | t <- axis1', t' <- axis2']
+        rngPairs = pairs axis1 axis2
+
+-- | 1-dimensional radial basis function kernel
+kernel1d_rbf :: Double -> Double -> Tensor -> Tensor -> Tensor
+kernel1d_rbf sigma length t t' = (sigma'^2) * exp eterm
+    where
+        sigma' = asTensor sigma
+        eterm = cmul (- (pow (t - t') (2 :: Int))) (1 / 2 * length^2)
+
+-- | derive a covariance matrix from the kernel for points on the axis
+makeCovmatrix :: [Float] -> [Float] -> Tensor
+makeCovmatrix axis1 axis2 = 
+    reshape (kernel1d_rbf 1.0 1.0 t t') [length axis1, length axis2]
+    where
+      (t, t') = makeAxis axis1 axis2
+
+-- | Multivariate 0-mean normal via cholesky decomposition
+mvnCholesky :: Tensor -> Int -> Int -> IO Tensor
+mvnCholesky cov axisDim n = do
+    samples <- randn' [axisDim, n]
+    pure $ matmul l samples
+    where 
+      l = cholesky cov Upper
+
+-- | Compute posterior mean and covariance parameters based on observed data y
+condition :: Tensor -> Tensor -> Tensor -> Tensor -> Tensor -> Tensor -> (Tensor, Tensor)
+condition muX muY covXX covXY covYY y =
+    (postMu, postCov)
+    where
+        covYX = transpose2D covXY
+        invY = inverse covYY
+        postMu = muX + (matmul covXY (matmul invY (y - muY)))
+        postCov = covXX - (matmul covXY (matmul invY covYX))
```

Comment:
The precision of `gesv` may be better than one of `inverse`.
The code becomes a bit complicated.

Reply:
Do you know what those map to in the libtorch API? The raw lapack library isn't surfaced in the C++ API. Searching the pytorch repo, the closest I can find is https://github.com/pytorch/pytorch/blob/054c7eb0f4e854e764f2dd4aa727fbb575ee02b8/aten/src/ATen/native/BatchLinearAlgebra.cpp that doesn't seem to correspond to functions surfaced in the spec.

This not being a full-blown gp library yet, i can accept an okay-ish solution if it keeps the code and library usage simpler.

Reply:
Sorry for not checking those map. (PR is acceptable.)
Those map is something complicated, and `posv` does not exist.

libtorch | lapack/blas
---------|---
gels     | gels
solve    | [sd]gesv
triangular_solve | [sd]trtrs
inverse | [sd]getri
cholesky_solve | [sd]potrs
cholesky | [sd]potrf
pstrf | [sd]pstrf
qr | qr
geqrf | geqrf
orgqr | orgqr
ormqr | ormqr